### PR TITLE
vmagent: add match_first_network support for docker sd

### PR DESCRIPTION
**Pull request by jiekun**
### Describe Your Changes
https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7398

add `match_first_network` support for docker service discovery. 
It uses the first network if the container has multiple networks defined, thus avoiding collecting duplicate targets.

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

File: lib/promscrape/discovery/docker/container.go [hunk 1]
```diff
@@ -67,14 +69,51 @@ func parseContainers(data []byte) ([]container, error) {
...
 	for i := range containers {
 		c := &containers[i]
 		if len(c.Names) == 0 {
 			continue
 		}
-		for _, n := range c.NetworkSettings.Networks {
+
+		networks := c.NetworkSettings.Networks
+		networkMode := c.HostConfig.NetworkMode
+		if len(networks) == 0 {
```
Comment (f41gh7) [inline lib/promscrape/discovery/docker/container.go:87]: I think we should put some love to this code. Currently, it looks fragile. I don't really like this endless for loop.

 First of all, I'd suggest to keep previous behavior and alternate it under this condition:

```
if matchFirstNetwork && len(networks) == 0
  // Try to lookup shared networks
}
```

File: lib/promscrape/discovery/docker/container.go [hunk 2]
```diff
@@ -67,14 +69,51 @@ func parseContainers(data []byte) ([]container, error) {
 	return containers, nil
 }
 
-func addContainersLabels(containers []container, networkLabels map[string]*promutils.Labels, defaultPort int, hostNetworkingHost string) []*promutils.Labels {
+func addContainersLabels(containers []container, networkLabels map[string]*promutils.Labels, defaultPort int, hostNetworkingHost string, matchFirstNetwork bool) []*promutils.Labels {
+	allContainers := make(map[string]container)
```
Comment (f41gh7) [inline lib/promscrape/discovery/docker/container.go:73]: I'd suggest to rename it for `containersByID`

Comment (f41gh7) [reply lib/promscrape/discovery/docker/container.go:73]: And use `map[string]*container` for storing results:

```go
var containersByID map[string]*container
if matchFirstNetwork{
  containersByID = make(map[string]*container)
  for _, c := range containers {
		containersByID[c.ID] = c
  }
}
```

File: lib/promscrape/discovery/docker/container.go [hunk 3]
```diff
@@ -67,14 +69,51 @@ func parseContainers(data []byte) ([]container, error) {
...
 		if len(c.Names) == 0 {
 			continue
 		}
-		for _, n := range c.NetworkSettings.Networks {
+
+		networks := c.NetworkSettings.Networks
+		networkMode := c.HostConfig.NetworkMode
+		if len(networks) == 0 {
+			// Try to lookup shared networks
+			for {
```
Comment (f41gh7) [inline lib/promscrape/discovery/docker/container.go:89]: I tried to understand purpose of this `for` loop and it seems not needed to me. 

We can just check the reference at `connected` container:

```
    containerID, ok := containerID(networkMode)
    if ok{
      tmpContainer, exist := allContainers[containerID]
      if exist{
        networks = tmpContainer.NetworkSettings.Networks    
      }
    }
```

Comment (jiekun) [reply lib/promscrape/discovery/docker/container.go:89]: It's possible that the connected container will connect to another container.
```json
[{
	"container_id": "1",
	"network": "some network"
}, {
	"container_id": "2",
	"network": "container:1"
}, {
	"container_id": "3",
	"network": "container:2"
}]
```

File: lib/promscrape/discovery/docker/container.go (hunk 2 above)
Comment (jiekun) [reply lib/promscrape/discovery/docker/container.go:73]: But one container could contains multiple ports. For example, if we have networks and container like this:
```json
[{
		"name": "network_1",
		"id": "...",
		"ip": "192.168.0.1"
	},
	{
		"name": "network_2 ",
		"id": "...",
		"ip": "192.168.0.2"
	}
]
```

```[{
		"name": "container_1",
		"ports": [80, 8080],
		"network": ["network_1", "network_2"]
	},
	{
		"name": "container_2",
		"ports": [90, 9090],
		"network": ["network_1", "network_2"]
	}
]
```

We will have 8 scrape targets:
```
192.168.0.1:80  // c1 + p1 + n1
192.168.0.1:8080  // c1 + p2 + n1
192.168.0.2:80  // c1 + p1 + n2
192.168.0.2:8080  // c1 + p2 + n2
192.168.0.1:90  // ...
192.168.0.1:9090 // ...
192.168.0.2:90 // ...
192.168.0.2:9090 // ...
```

**Comment by jiekun** [comment]
@f41gh7 I've seperated the changelog into 2 entries. PTAL.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**